### PR TITLE
SLD package extension [GEOS-6837]

### PR DESCRIPTION
**Pull request by josegar74**
A extension to upload with the REST API a zip package with SLD file and related resources (images) to Geoserver. 

Code to review. 

https://jira.codehaus.org/browse/GEOS-6837

### CODE REVIEW

**Comment by jodygarnett** [comment]
Without an update this won't make it into the GeoServer 2.7 release...

**Comment by mbarto** [comment]
I can try to review the PR and add some tests tomorrow, so that it can be merged before the freeze.

**Comment by mbarto** [comment]
Hi created a new pull request for this  (https://github.com/geoserver/geoserver/pull/894) with some tests and basic documentation.
I also created a JIRA (http://jira.codehaus.org/browse/GEOS-6837) for it.

**Comment by josegar74** [comment]
Hi

Apologies, as I have been busy with other stuff and could not manage about this. 

I started with Justin suggestions, adding some unit testing and refactoring the code in this branch: https://github.com/josegar74/geoserver/commits/sld_package_styleresource , but requires still some checks.

Looks like @mbarto looks has done some improvements of the original code, so I guess can be fine.

**Comment by aaime** [comment]
See the thread on geoserver-devel, it appears that Mauro's work does not yet match Justin's original request (it still creates a new url to hit)

**Comment by josegar74** [comment]
I see, I'll check next week to find some time to finish this. Apologies for delays in providing it.

**Comment by jdeolive** [comment]
Fwiw I don't want to have my feedback be a blocker, if there is a big push
to get it in now as is go for it.

**Comment by aaime** [comment]
Eh, at this point I believe we might be too late, Jody is going to cut 2.7-beta Monday and the feature freeze will start.

**Comment by jodygarnett** [comment]
I also noted the new "sld" endpoint and found it a but rough (especially as
we now support multiple styles for css). I am happy to wait until after the
code freeze and back port this REST API change when it is ready.

--
Jody Garnett

File: src/restconfig/src/main/java/applicationContext.xml [hunk 1]
```diff
@@ -512,11 +512,17 @@
             <key><value>/workspaces/{workspace}/coveragestores/{coveragestore}/coverages/{coverage}/index/granules/{granule}</value></key>
             <value>structuredCoverageFinder</value>
         </entry>
+        <entry>
+          <key><value>/sld/{workspace}/{style}</value></key>
+          <value>stylePackageFinder</value>
+        </entry>
```
Comment (jodygarnett) [inline src/restconfig/src/main/java/applicationContext.xml]: I assume this is the new endpoint, can we do anything more generic (as sld is not the only thing in the mix - see StyleHandler interface).

File: src/restconfig/src/main/java/org/geoserver/catalog/rest/StylePackageResource.java [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+     */
+    private Style parseSld(File sldFile) {
+        Style style = null;
+        InputStream is = null;
+
+        try {
+            is = new FileInputStream(sldFile);
+
+            SLDParser parser
+                    = new SLDParser(CommonFactoryFinder.getStyleFactory(null), is);
```
Comment (jodygarnett) [inline src/restconfig/src/main/java/org/geoserver/catalog/rest/StylePackageResource.java]: I feel like I reviewed this code snippet before - do we want to try supporting several parsers here. Or use StyleHandler to support different kinds of files?

**Comment by jodygarnett** [comment]
Looks like we should move conversation over to #894 and mark this request as closed?

**Comment by josegar74** [comment]
Updated the code to use the the style resource url, added unit tests and documentation.

File: doc/en/user/source/rest/api/styles.rst [hunk 3]
```diff
@@ -94,7 +95,7 @@ Controls a given style.
    * - GET
      - Return style ``s``
      - 200
-     - SLD, HTML, XML, JSON
+     - SLD, HTML, XML, JSON, ZIP
      - HTML
```
Comment (mbarto) [inline doc/en/user/source/rest/api/styles.rst]: I don't think ZIP (SLDPackage) is supported for GET: am I wrong?

File: src/main/src/main/java/org/geoserver/catalog/SLDPackageHandler.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
+/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
+ * (c) 2001 - 2014 OpenPlans
```
Comment (mbarto) [inline src/main/src/main/java/org/geoserver/catalog/SLDPackageHandler.java]: This line can be removed: it's a new file

File: doc/en/user/source/rest/api/styles.rst (hunk 3 above)
Comment (josegar74) [inline doc/en/user/source/rest/api/styles.rst]: The SLDPackageHandler.encode manages about this. But maybe is better to just delegate to sldHandler.encode and return the SLD raw file. What do you think?

Comment (mbarto) [inline doc/en/user/source/rest/api/styles.rst]: Not sure, is the GET with application/zip content-type just zipping the sld file? We don't have the images in the file I suppose. Probably it's not worth having this if we are not creating a "package" for GET too.

Comment (josegar74) [inline doc/en/user/source/rest/api/styles.rst]: For now yes, only the sld. Changed to return the SLD instead. In the future can be improved to return a full package with all related resources if that is useful.

**Comment by mbarto** [comment]
Looks good to me with latest changes.

**Comment by aaime** [comment]
Nice. We just need to wait the freeze to be over and then we can merge on master (GeoServer 2.8.x)

**Comment by jodygarnett** [comment]
I think the feature freeze is over, there is now a master and a 13.x branch
(we only have build problems with the 13.x jobs).

--
Jody Garnett

**Comment by aaime** [comment]
You mean a 2.7.x branch: https://github.com/geoserver/geoserver/tree/2.7.x ;-)

**Comment by aaime** [comment]
All right, let's merge then

**Pull request merged**